# Домашнее задание: cравнение древовидных алгоритмов на NLP-данных

**Задача:** oбучить разные древовидные модели на текстах новостей, замерить test accuracy и время обучения, заполнить таблицы.

## 1. Загрузка и подготовка данных (код дан)

In [ ]:
from sklearn.datasets import fetch_20newsgroups
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score
import time
import pandas as pd

# Загружаем 3 категории
categories = ['rec.sport.baseball', 'sci.space', 'comp.graphics']
newsgroups = fetch_20newsgroups(subset='all', categories=categories, shuffle=True, random_state=42)

# TF-IDF векторизация
vectorizer = TfidfVectorizer(max_features=5000, ngram_range=(1, 2), stop_words='english')
X = vectorizer.fit_transform(newsgroups.data)
y = newsgroups.target

# Разделение на train/test
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

print(f"Размер обучающей выборки: {X_train.shape}")
print(f"Размер тестовой выборки: {X_test.shape}")

Размер обучающей выборки: (2363, 5000)
Размер тестовой выборки: (591, 5000)


## 2. Образец: Дерево решений (Decision Tree)

**Гиперпараметр:** `max_depth`

In [ ]:
from sklearn.tree import DecisionTreeClassifier

results_tree = []

# Образец: мы делаем перебор значений max_depth = [3, 5, 10, 20, None])
# Для каждого: замер времени, обучение, accuracy на тесте, сохранение в results_tree

for depth in [3, 5, 10, 20, None]:
    start = time.time()
    clf = DecisionTreeClassifier(max_depth=depth, random_state=42)
    clf.fit(X_train, y_train)
    fit_time = time.time() - start
    acc = accuracy_score(y_test, clf.predict(X_test))
    results_tree.append({'max_depth': depth, 'test_accuracy': round(acc, 4), 'time_sec': round(fit_time, 2)})
    print(f"depth={depth}: acc={acc:.4f}, time={fit_time:.2f}s")

df_tree = pd.DataFrame(results_tree)
print("\nТаблица 1. Дерево решений")
print(df_tree.to_markdown(index=False))

depth=3: acc=0.6244, time=0.12s
depth=5: acc=0.6971, time=0.18s
depth=10: acc=0.7716, time=0.35s
depth=20: acc=0.8409, time=0.47s
depth=None: acc=0.8697, time=0.60s

Таблица 1. Дерево решений
|   max_depth |   test_accuracy |   time_sec |
|------------:|----------------:|-----------:|
|           3 |          0.6244 |       0.12 |
|           5 |          0.6971 |       0.18 |
|          10 |          0.7716 |       0.35 |
|          20 |          0.8409 |       0.47 |
|         nan |          0.8697 |       0.6  |


## 3. Заполните таблицы для остальных алгоритмов

По аналогии с образцом обучите следующие модели и запишите результаты.

### 3.1. Случайный лес (Random Forest)

**Гиперпараметр:** `max_depth` (фиксируем `n_estimators=100`)

| max_depth | test_accuracy | time_sec |
|-----------|---------------|----------|
| 5         |               |          |
| 10        |               |          |
| 20        |               |          |
| None      |               |          |

In [ ]:
from sklearn.ensemble import RandomForestClassifier

results_rf = []

# Ваш код здесь (перебор max_depth)
for depth in [5, 10, 20, None]:
    start = time.time()
    clf = RandomForestClassifier(n_estimators=100, max_depth=depth, random_state=42)
    clf.fit(X_train, y_train)
    fit_time = time.time() - start
    acc = accuracy_score(y_test, clf.predict(X_test))
    results_rf.append({'max_depth': depth, 'test_accuracy': round(acc, 4), 'time_sec': round(fit_time, 2)})
    print(f"depth={depth}: acc={acc:.4f}, time={fit_time:.2f}s")

df_rf = pd.DataFrame(results_rf)
print("\nТаблица 2. Случайный лес")
print(df_rf.to_markdown(index=False))

depth=5: acc=0.9002, time=1.48s
depth=10: acc=0.9154, time=1.61s
depth=20: acc=0.9357, time=1.70s
depth=None: acc=0.9526, time=3.17s

Таблица 2. Случайный лес
|   max_depth |   test_accuracy |   time_sec |
|------------:|----------------:|-----------:|
|           5 |          0.9002 |       1.48 |
|          10 |          0.9154 |       1.61 |
|          20 |          0.9357 |       1.7  |
|         nan |          0.9526 |       3.17 |


### 3.2. Градиентный бустинг (Gradient Boosting)

**Гиперпараметр:** `learning_rate` (фиксируем `n_estimators=100, max_depth=3`)

| learning_rate | test_accuracy | time_sec |
|---------------|---------------|----------|
| 0.01          |               |          |
| 0.05          |               |          |
| 0.1           |               |          |
| 0.5           |               |          |

In [ ]:
from sklearn.ensemble import GradientBoostingClassifier

results_gb = []

# Ваш код здесь (перебор learning_rate)
for lr in [0.01, 0.05, 0.1, 0.5]:
    start = time.time()
    clf = GradientBoostingClassifier(n_estimators=100, max_depth=3, learning_rate=lr, random_state=42)
    clf.fit(X_train, y_train)
    fit_time = time.time() - start
    acc = accuracy_score(y_test, clf.predict(X_test))
    results_gb.append({'learning_rate': lr, 'test_accuracy': round(acc, 4), 'time_sec': round(fit_time, 2)})
    print(f"learning_rate={lr}: acc={acc:.4f}, time={fit_time:.2f}s")

df_gb = pd.DataFrame(results_gb)
print("\nТаблица 3. Градиентный бустинг")
print(df_gb.to_markdown(index=False))

learning_rate=0.01: acc=0.8782, time=44.96s
learning_rate=0.05: acc=0.9475, time=35.59s
learning_rate=0.1: acc=0.9577, time=37.33s
learning_rate=0.5: acc=0.9662, time=35.88s

Таблица 3. Градиентный бустинг
|   learning_rate |   test_accuracy |   time_sec |
|----------------:|----------------:|-----------:|
|            0.01 |          0.8782 |      44.96 |
|            0.05 |          0.9475 |      35.59 |
|            0.1  |          0.9577 |      37.33 |
|            0.5  |          0.9662 |      35.88 |


### 3.3. XGBoost

**Гиперпараметр:** `gamma` (фиксируем `n_estimators=100, learning_rate=0.1`)

| gamma | test_accuracy | time_sec |
|-------|---------------|----------|
| 0     |               |          |
| 0.1   |               |          |
| 0.5   |               |          |
| 1.0   |               |          |

In [ ]:
from xgboost import XGBClassifier

results_xgb = []

# Ваш код здесь (перебор gamma, use_label_encoder=False, eval_metric='logloss')
for gamma_val in [0, 0.1, 0.5, 1.0]:
    start = time.time()
    clf = XGBClassifier(n_estimators=100, learning_rate=0.1, gamma=gamma_val, random_state=42, use_label_encoder=False, eval_metric='logloss')
    clf.fit(X_train, y_train)
    fit_time = time.time() - start
    acc = accuracy_score(y_test, clf.predict(X_test))
    results_xgb.append({'gamma': gamma_val, 'test_accuracy': round(acc, 4), 'time_sec': round(fit_time, 2)})
    print(f"gamma={gamma_val}: acc={acc:.4f}, time={fit_time:.2f}s")

df_xgb = pd.DataFrame(results_xgb)
print("\nТаблица 4. XGBoost")
print(df_xgb.to_markdown(index=False))

/usr/local/lib/python3.12/dist-packages/xgboost/training.py:200: UserWarning: [16:24:43] WARNING: /__w/xgboost/xgboost/src/learner.cc:782: 
Parameters: { "use_label_encoder" } are not used.

  bst.update(dtrain, iteration=i, fobj=obj)


gamma=0: acc=0.9459, time=38.11s


/usr/local/lib/python3.12/dist-packages/xgboost/training.py:200: UserWarning: [16:25:21] WARNING: /__w/xgboost/xgboost/src/learner.cc:782: 
Parameters: { "use_label_encoder" } are not used.

  bst.update(dtrain, iteration=i, fobj=obj)


gamma=0.1: acc=0.9442, time=34.00s


/usr/local/lib/python3.12/dist-packages/xgboost/training.py:200: UserWarning: [16:25:55] WARNING: /__w/xgboost/xgboost/src/learner.cc:782: 
Parameters: { "use_label_encoder" } are not used.

  bst.update(dtrain, iteration=i, fobj=obj)


gamma=0.5: acc=0.9492, time=30.54s


/usr/local/lib/python3.12/dist-packages/xgboost/training.py:200: UserWarning: [16:26:25] WARNING: /__w/xgboost/xgboost/src/learner.cc:782: 
Parameters: { "use_label_encoder" } are not used.

  bst.update(dtrain, iteration=i, fobj=obj)


gamma=1.0: acc=0.9492, time=30.82s

Таблица 4. XGBoost
|   gamma |   test_accuracy |   time_sec |
|--------:|----------------:|-----------:|
|     0   |          0.9459 |      38.11 |
|     0.1 |          0.9442 |      34    |
|     0.5 |          0.9492 |      30.54 |
|     1   |          0.9492 |      30.82 |


### 3.4. AdaBoost

**Гиперпараметр:** `learning_rate` (фиксируем `n_estimators=100`)

| learning_rate | test_accuracy | time_sec |
|---------------|---------------|----------|
| 0.5           |               |          |
| 1.0           |               |          |
| 1.5           |               |          |
| 2.0           |               |          |

In [ ]:
from sklearn.ensemble import AdaBoostClassifier

results_ada = []

# Ваш код здесь (перебор learning_rate)
for lr in [0.5, 1.0, 1.5, 2.0]:
    start = time.time()
    clf = AdaBoostClassifier(n_estimators=100, learning_rate=lr, random_state=42)
    clf.fit(X_train, y_train)
    fit_time = time.time() - start
    acc = accuracy_score(y_test, clf.predict(X_test))
    results_ada.append({'learning_rate': lr, 'test_accuracy': round(acc, 4), 'time_sec': round(fit_time, 2)})
    print(f"learning_rate={lr}: acc={acc:.4f}, time={fit_time:.2f}s")

df_ada = pd.DataFrame(results_ada)
print("\nТаблица 5. AdaBoost")
print(df_ada.to_markdown(index=False))

learning_rate=0.5: acc=0.8629, time=9.77s
learning_rate=1.0: acc=0.9205, time=5.46s
learning_rate=1.5: acc=0.9255, time=5.59s
learning_rate=2.0: acc=0.8731, time=5.18s

Таблица 5. AdaBoost
|   learning_rate |   test_accuracy |   time_sec |
|----------------:|----------------:|-----------:|
|             0.5 |          0.8629 |       9.77 |
|             1   |          0.9205 |       5.46 |
|             1.5 |          0.9255 |       5.59 |
|             2   |          0.8731 |       5.18 |


### 3.5. LightGBM

**Гиперпараметр:** `num_leaves` (фиксируем `n_estimators=100, learning_rate=0.1`)

| num_leaves | test_accuracy | time_sec |
|------------|---------------|----------|
| 15         |               |          |
| 31         |               |          |
| 63         |               |          |
| 127        |               |          |

In [ ]:
from lightgbm import LGBMClassifier

results_lgbm = []

# Ваш код здесь (перебор num_leaves, n_jobs=-1)
for leaves in [15, 31, 63, 127]:
    start = time.time()
    clf = LGBMClassifier(n_estimators=100, learning_rate=0.1, num_leaves=leaves, random_state=42, n_jobs=-1)
    clf.fit(X_train, y_train)
    fit_time = time.time() - start
    acc = accuracy_score(y_test, clf.predict(X_test))
    results_lgbm.append({'num_leaves': leaves, 'test_accuracy': round(acc, 4), 'time_sec': round(fit_time, 2)})
    print(f"num_leaves={leaves}: acc={acc:.4f}, time={fit_time:.2f}s")

df_lgbm = pd.DataFrame(results_lgbm)
print("\nТаблица 6. LightGBM")
print(df_lgbm.to_markdown(index=False))

[LightGBM] [Info] Auto-choosing row-wise multi-threading, the overhead of testing was 0.062606 seconds.
You can set `force_row_wise=true` to remove the overhead.
And if memory is not enough, you can set `force_col_wise=true`.
[LightGBM] [Info] Total Bins 61063
[LightGBM] [Info] Number of data points in the train set: 2363, number of used features: 2963
[LightGBM] [Info] Start training from score -1.119999
[LightGBM] [Info] Start training from score -1.093126
[LightGBM] [Info] Start training from score -1.083076


/usr/local/lib/python3.12/dist-packages/sklearn/utils/validation.py:2739: UserWarning: X does not have valid feature names, but LGBMClassifier was fitted with feature names
  warnings.warn(


num_leaves=15: acc=0.9662, time=4.09s
[LightGBM] [Info] Auto-choosing col-wise multi-threading, the overhead of testing was 0.069653 seconds.
You can set `force_col_wise=true` to remove the overhead.
[LightGBM] [Info] Total Bins 61063
[LightGBM] [Info] Number of data points in the train set: 2363, number of used features: 2963
[LightGBM] [Info] Start training from score -1.119999
[LightGBM] [Info] Start training from score -1.093126
[LightGBM] [Info] Start training from score -1.083076


/usr/local/lib/python3.12/dist-packages/sklearn/utils/validation.py:2739: UserWarning: X does not have valid feature names, but LGBMClassifier was fitted with feature names
  warnings.warn(


num_leaves=31: acc=0.9594, time=9.23s
[LightGBM] [Info] Auto-choosing row-wise multi-threading, the overhead of testing was 0.074925 seconds.
You can set `force_row_wise=true` to remove the overhead.
And if memory is not enough, you can set `force_col_wise=true`.
[LightGBM] [Info] Total Bins 61063
[LightGBM] [Info] Number of data points in the train set: 2363, number of used features: 2963
[LightGBM] [Info] Start training from score -1.119999
[LightGBM] [Info] Start training from score -1.093126
[LightGBM] [Info] Start training from score -1.083076
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf


/usr/local/lib/python3.12/dist-packages/sklearn/utils/validation.py:2739: UserWarning: X does not have valid feature names, but LGBMClassifier was fitted with feature names
  warnings.warn(


num_leaves=63: acc=0.9577, time=13.03s
[LightGBM] [Info] Auto-choosing row-wise multi-threading, the overhead of testing was 0.068064 seconds.
You can set `force_row_wise=true` to remove the overhead.
And if memory is not enough, you can set `force_col_wise=true`.
[LightGBM] [Info] Total Bins 61063
[LightGBM] [Info] Number of data points in the train set: 2363, number of used features: 2963
[LightGBM] [Info] Start training from score -1.119999
[LightGBM] [Info] Start training from score -1.093126
[LightGBM] [Info] Start training from score -1.083076
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain:

/usr/local/lib/python3.12/dist-packages/sklearn/utils/validation.py:2739: UserWarning: X does not have valid feature names, but LGBMClassifier was fitted with feature names
  warnings.warn(


### 3.6. CatBoost

**Гиперпараметр:** `depth` (фиксируем `iterations=100, learning_rate=0.1`)

| depth | test_accuracy | time_sec |
|-------|---------------|----------|
| 3     |               |          |
| 5     |               |          |
| 7     |               |          |
| 10    |               |          |

In [ ]:
!pip install catboost
from catboost import CatBoostClassifier

results_cat = []

# Ваш код здесь (перебор depth, verbose=0)

for depth_val in [3, 5, 7, 10]:
    start = time.time()
    clf = CatBoostClassifier(iterations=100, learning_rate=0.1, depth=depth_val, random_state=42, verbose=0)
    clf.fit(X_train, y_train)
    fit_time = time.time() - start
    acc = accuracy_score(y_test, clf.predict(X_test))
    results_cat.append({'depth': depth_val, 'test_accuracy': round(acc, 4), 'time_sec': round(fit_time, 2)})
    print(f"depth={depth_val}: acc={acc:.4f}, time={fit_time:.2f}s")

df_cat = pd.DataFrame(results_cat)
print("\nТаблица 7. CatBoost")
print(df_cat.to_markdown(index=False))

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 97.1/97.1 MB 8.9 MB/s eta 0:00:00
depth=3: acc=0.9340, time=11.96s
depth=5: acc=0.9425, time=33.65s
depth=7: acc=0.9442, time=112.54s
depth=10: acc=0.9526, time=688.15s

Таблица 7. CatBoost
|   depth |   test_accuracy |   time_sec |
|--------:|----------------:|-----------:|
|       3 |          0.934  |      11.96 |
|       5 |          0.9425 |      33.65 |
|       7 |          0.9442 |     112.54 |
|      10 |          0.9526 |     688.15 |


## 4. Итоговая таблица лучших результатов

In [ ]:
# Соберите лучшие результаты из каждой таблицы вручную или автоматически
# Пример структуры:

summary = pd.DataFrame([
    ('Decision Tree', 0, 0, 0),  # замените 0 на ваши значения
    ('Random Forest', 0, 0, 0),
    ('Gradient Boosting', 0, 0, 0),
    ('XGBoost', 0, 0, 0),
    ('AdaBoost', 0, 0, 0),
    ('LightGBM', 0, 0, 0),
    ('CatBoost', 0, 0, 0),
], columns=['Algorithm', 'Best params', 'Best test acc', 'Time (sec)'])

print("Итоговая таблица лучших результатов")
print(summary.to_markdown(index=False))

In [ ]:
best_tree = max(results_tree, key=lambda x: x['test_accuracy'])
best_rf = max(results_rf, key=lambda x: x['test_accuracy'])
best_gb = max(results_gb, key=lambda x: x['test_accuracy'])
best_xgb = max(results_xgb, key=lambda x: x['test_accuracy'])
best_ada = max(results_ada, key=lambda x: x['test_accuracy'])
best_lgbm = max(results_lgbm, key=lambda x: x['test_accuracy'])
best_cat = max(results_cat, key=lambda x: x['test_accuracy'])

summary = pd.DataFrame([
    ('Decision Tree', best_tree['max_depth'], best_tree['test_accuracy'], best_tree['time_sec']),
    ('Random Forest', best_rf['max_depth'], best_rf['test_accuracy'], best_rf['time_sec']),
    ('Gradient Boosting', best_gb['learning_rate'], best_gb['test_accuracy'], best_gb['time_sec']),
    ('XGBoost', best_xgb['gamma'], best_xgb['test_accuracy'], best_xgb['time_sec']),
    ('AdaBoost', best_ada['learning_rate'], best_ada['test_accuracy'], best_ada['time_sec']),
    ('LightGBM', best_lgbm['num_leaves'], best_lgbm['test_accuracy'], best_lgbm['time_sec']),
    ('CatBoost', best_cat['depth'], best_cat['test_accuracy'], best_cat['time_sec']),
], columns=['Algorithm', 'Best params', 'Best test acc', 'Time (sec)'])

print("Итоговая таблица лучших результатов")
print(summary.to_markdown(index=False))

Итоговая таблица лучших результатов
| Algorithm         |   Best params |   Best test acc |   Time (sec) |
|:------------------|--------------:|----------------:|-------------:|
| Decision Tree     |         nan   |          0.8697 |         0.6  |
| Random Forest     |         nan   |          0.9526 |         3.17 |
| Gradient Boosting |           0.5 |          0.9662 |        35.88 |
| XGBoost           |           0.5 |          0.9492 |        30.54 |
| AdaBoost          |           1.5 |          0.9255 |         5.59 |
| LightGBM          |          15   |          0.9662 |         4.09 |
| CatBoost          |          10   |          0.9526 |       688.15 |


## 5. Вопросы (ответить текстом в следующей ячейке)

1. Какой алгоритм показал максимальную точность? Какие параметры к этому привели?

2. Какой алгоритм быстрее всего обучался? Во сколько раз он быстрее самого медленного?

3. У каких алгоритмов наблюдалось переобучение? При каких параметрах?

4. Какой алгоритм вы выбрали бы для продакшн-системы с миллионом текстов? Почему?

# Напишите свои ответы здесь

**Ответ 1:** Максимальную точность (0,9662) показали Gradient Boosting с learning_rate=0,5 и LightGBM с num_leaves=15. Оба алгоритма правильно классифицировали более 96% текстов тестовой выборки.


**Ответ 2:** Быстрее всех обучалось Decision Tree — 0,6 секунды. CatBoost при depth=10 оказался самым медленным (688 секунд). Такая значительная разница объясняется тем, что CatBoost строит симметричные деревья и на каждой итерации перебирает все разбиения.


**Ответ 3:** Переобучение явно наблюдалось у AdaBoost при learning_rate=2,0: точность упала с 92,5% (при learning_rate=1,5) до 87,3% из-за слишком большого шага градиента. У LightGBM при num_leaves=63 и 127 также было небольшое переобучение — точность снизилась на 0,7% относительно лучшего результата, хотя время обучения выросло в 3-4 раза.


**Ответ 4:** Для продакшн-системы с миллионом текстов я бы выбрала LightGBM. При той же максимальной точности, что и у Gradient Boosting (96,6%), LightGBM обучается в 8 раз быстрее (4 секунды против 36) и занимает значительно меньше памяти благодаря гистограммному методу и работе с разреженными матрицами. CatBoost показывает сопоставимую точность, но требует 11,5 минут обучения.

## Критерии оценки

| Что оценивается | Баллы |
|----------------|-------|
| Заполнены все 7 таблиц (правильно собраны accuracy и время) | 3 |
| Заполнена итоговая таблица лучших результатов | 1 |
| Ответы на 4 вопроса (по 0.5 балла) | 2 |
| Код воспроизводим (random_state=42, порядок ячеек корректен) | 1 |
| **Качество кода** (отсутствие дублирования, осмысленные имена переменных, циклы вместо копипасты, использование .to_markdown()) | 3 |
| **Итого** | **10** |

### Детали по качеству кода (+3 балла):
- **+1** — использование единого шаблона для всех экспериментов (цикл + list.append + pd.DataFrame)
- **+1** — правильное форматирование вывода (округление времени и accuracy до 2-4 знаков)
- **+1** — читаемые имена переменных, комментарии, отсутствие `eval()`, `exec()` и магических чисел

**Штрафы:**
- -1 балл, если код не запускается без ошибок
- -1 балл, если отсутствует `random_state=42` в любой из моделей
- -1 балл, если таблицы выведены криво (не через `to_markdown` или нечитаемый `print`)